File: algorithms/graph/planarity-testing/python/planarity_testing.py
def planarity_testing(arr: list[int]) -> int:
    n = arr[0]
    m = arr[1]
    if n <= 4 and m <= 6:
        # For very small graphs, count unique edges
        edges = set()
        for i in range(m):
            u = arr[2 + 2 * i]
            v = arr[2 + 2 * i + 1]
            if u != v:
                edges.add((min(u, v), max(u, v)))
        e = len(edges)
        if n < 3:
            return 1
        return 1 if e <= 3 * n - 6 else 0

    edges = set()
    for i in range(m):
        u = arr[2 + 2 * i]
        v = arr[2 + 2 * i + 1]
        if u != v:
            edges.add((min(u, v), max(u, v)))
    e = len(edges)

    if n < 3:
        return 1
    if e > 3 * n - 6:
        return 0
    return 1

File: algorithms/graph/planarity-testing/go/planarity_testing.go
package planaritytesting

func PlanarityTesting(arr []int) int {
	n := arr[0]; m := arr[1]
	type edge struct{ a, b int }
	edges := make(map[edge]bool)
	for i := 0; i < m; i++ {
		u, v := arr[2+2*i], arr[2+2*i+1]
		if u != v {
			a, b := u, v
			if a > b { a, b = b, a }
			edges[edge{a, b}] = true
		}
	}
	e := len(edges)
	if n < 3 { return 1 }
	if e <= 3*n-6 { return 1 }
	return 0
}
